kbchat/kbchat_test.go

```diff
@@ -0,0 +1,117 @@
+package kbchat
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v2"
+)
+
+type Bots struct {
+	Alice1   *OneshotOptions
+	Alice2   *OneshotOptions
+	Bob1     *OneshotOptions
+	Charlie1 *OneshotOptions
+}
+
+type Config struct {
+	Bots
+}
+
+func readAndParseConfig() (Config, error) {
+	var config Config
+	data, err := ioutil.ReadFile("test_config.yaml")
+	if err != nil {
+		return Config{}, err
+	}
+
+	err = yaml.Unmarshal(data, &config)
+	if err != nil {
+		return Config{}, err
+	}
+
+	return config, nil
+}
+
+func randomTempDir() (string, error) {
+	bytes := make([]byte, 16)
+	_, err := rand.Read(bytes)
+	if err != nil {
+		return "", err
+	}
+	dir := path.Join(os.TempDir(), "keybase_bot_"+hex.EncodeToString(bytes))
+	return dir, nil
+}
+
+func whichKeybase() (string, error) {
+	cmd := exec.Command("which", "keybase")
+
+	out, err := cmd.Output()
+	if err != nil {
+		return "", err
+	}
+	location := strings.TrimSpace(string(out))
+	return location, nil
+}
+
+func copyFile(source, dest string) error {
+	sourceData, err := ioutil.ReadFile(source)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(dest, sourceData, 0777)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func prepWorkingDir(workingDir string) (string, error) {
```

Comment:
Just was looking at this again, why do you need to prep the working directory? The keybase binary doesn't need to be inside of the home directory and at least in my experience `keybase --home /tmp/keyboardmash service` works fine and will automatically create `/tmp/keyboardmash` and everything it needs inside of it. 

Which I think means you could also get rid of `copyFile`, `whichKeybase`, and `whichKeybase`.

Reply:
I was getting an issue with the homedir arg length. (something like "`Chdir` cannot be over 108 characters") I think we ran into a similar issue with the TS bot, and the solution was to isolate the binary, which @malgorithms implemented (keybase/keybase-bot#114). That fixed the issue; I can't remember if there were some other motivations behind it too or not

Reply:
Huh interesting. Kinda surprised I never ran into that while working on the SSH bot but :+1:

Reply:
@ddworken - not sure if this is what you're asking, but .since a bot consists of multiple processes:

- a service which is supposed to run indefinitely, possibly weeks
- listener processes which may also do the same
- one-off client launches (for sending messages and things like that)

...then we have a possible version mismatch problem.  For example, we can't have a scenario where you start a bot script up, and then the next day your keybase app (which maybe you're also using personally) gets an update. If you're using that same binary, now you've got some listening processes on the old version (the service, any streaming listeners), but any one-off calls will be the new version. This could be a breaking change.

So the bot library simply makes its own binary available and isolates itself, so it never has any problems with uninstalls/upgrades/downtime/etc.

Dunno if this answers your question...

Reply:
Ah, yes that is exactly what I was wondering. Good point about version mismatches and this being a solution to that problem. @nathunsmitty maybe add a comment to the code explaining this logic?